src/sql/src/plan/statement/dml.rs

```diff
@@ -189,21 +191,57 @@ pub fn plan_select(
         return Ok(Plan::SideEffectingFunc(f));
     }
 
+    let (plan, _desc) = plan_select_inner(scx, select, params, copy_to)?;
+    Ok(Plan::Select(plan))
+}
+
+fn plan_select_inner(
+    scx: &StatementContext,
+    select: SelectStatement<Aug>,
+    params: &Params,
+    copy_to: Option<CopyFormat>,
+) -> Result<(SelectPlan, RelationDesc), PlanError> {
     let when = query::plan_as_of(scx, select.as_of)?;
     let query::PlannedRootQuery {
         mut expr,
-        desc: _,
+        desc,
         finishing,
         scope: _,
     } = query::plan_root_query(scx, select.query, QueryLifetime::OneShot)?;
     expr.bind_parameters(params)?;
 
-    Ok(Plan::Select(SelectPlan {
+    // A top-level limit cannot be data dependent so eagerly evaluate it.
+    let limit = match finishing.limit {
```

Comment:
Does this need to check the exprs in limit feature flag?

Reply:
The existing check already properly checks for all limits that are not simple constants!

https://github.com/MaterializeInc/materialize/blob/ccdcdb61709b485477fb0caf93304cc30cb3ceed/src/sql/src/plan/query.rs#L1301-L1304